src/AppContext.tsx

```diff
@@ -0,0 +1,292 @@
+  todos: Todo[],
+  status: Status
+  setTodos: (arg: Todo[]) => void,
+  setStatus: (arg: Status) => void,
+  deleteTodo: (arg: number) => void,
+  selectedTodoIds: number[],
+  setSelectedTodoIds: (arg: number[]) => void
+  handleToggleCompleted: (arg: Todo) => void,
+  clearCompleted: () => void,
+  todoTitle: string,
+  setTodoTitle: (arg: string) => void,
+  tempTodo: Todo | null,
+  setTempTodo: (arg: Todo | null) => void,
+  addTodo: (arg: string) => void,
+  error: ErrorType | null,
+  setError: (arg: ErrorType | null) => void,
+  loadind: boolean,
+  setLoading: (arg: boolean) => void,
+  updateTodo: (arg: Todo) => void,
+  handleToggleCompletedAll: () => void,
+}
+
+export const AppContext = createContext<AppContextType>({
+  todos: [],
+  status: Status.all,
+  setTodos: () => { },
+  setStatus: () => { },
+  deleteTodo: () => { },
+  selectedTodoIds: [],
+  setSelectedTodoIds: () => { },
+  handleToggleCompleted: () => { },
+  clearCompleted: () => { },
+  todoTitle: '',
+  setTodoTitle: () => { },
+  tempTodo: null,
+  setTempTodo: () => { },
+  addTodo: () => { },
+  error: null,
+  setError: () => { },
+  loadind: false,
+  setLoading: () => { },
+  updateTodo: () => { },
+  handleToggleCompletedAll: () => { },
+});
+
+type Props = {
+  children: React.ReactNode
+};
+
+export const AppProvider: React.FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [status, setStatus] = useState<Status>(Status.all);
+  const [selectedTodoIds, setSelectedTodoIds] = useState<number[]>([]);
+  const [todoTitle, setTodoTitle] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [error, setError] = useState<ErrorType | null>(null);
+  const [loadind, setLoading] = useState(false);
+
+  const deleteTodo = useCallback(
+    (todoId: number) => {
+      setSelectedTodoIds(currentIds => [...currentIds, todoId]);
+      todoService.deleteTodo(todoId)
+        .then(() => {
+          setTimeout(() => {
+            setTodos(currentTodos => currentTodos
+              .filter(post => post.id !== todoId));
+          }, 500);
+        })
+        .catch(() => {
+          setSelectedTodoIds(ids => {
+            ids.splice(ids.indexOf(todoId), 1);
+
+            return ids;
+          });
+          setError(ErrorType.cantDeleteTodo);
+          setTimeout(() => setError(null), 2000);
+        })
+        .finally(() => setTimeout(
+          () => setSelectedTodoIds(ids => ids.splice(ids.indexOf(todoId), 1)),
+          500,
+        ));
+    }, [],
+  );
+
+  const clearCompleted = useCallback(
+    () => {
+      const completedTodos = todos.filter(
+        todoToFind => todoToFind.completed,
+      );
+
+      setSelectedTodoIds(currentIds => (
+        [...currentIds, ...completedTodos
+          .map(completedTodo => completedTodo.id),
+        ]
+      ));
+
+      completedTodos.map(
+        completedTodo => todoService.deleteTodo(completedTodo.id),
+      );
```

Comment:
What if requests fail? You can wait for a response with `Promise.all` or `Promise.allSettled`